contracts/epochs/FlowEpoch.cdc

```diff
@@ -109,46 +121,511 @@ pub contract FlowEpochs {
+    pub var proposedEpochCounter: UInt64
+
+    /// The current phase that the epoch is in
+    pub var currentEpochPhase: EpochPhase
+
+    /// Contains a historical record of the metadata from all previous epochs
+    /// indexed by epoch number
+    access(contract) var epochMetadata: {UInt64: EpochMetadata}
+
+    access(contract) let QCAdmin: @FlowEpochClusterQC.Admin
+
+    access(contract) let DKGAdmin: @FlowDKG.Admin
+
+    access(contract) let stakingAdmin: @FlowIDTableStaking.Admin
+
+    pub let adminStoragePath: StoragePath
+    pub let heartbeatStoragePath: StoragePath
+
+    /// Resource that can update some of the contract fields
+    pub resource Admin {
+        pub fun updateEpochViews(_ newEpochViews: UInt64) {
+            pre {
+                FlowEpoch.currentEpochPhase == EpochPhase.STAKINGAUCTION: "Can only update fields during the staking auction"
+                newEpochViews > FlowEpoch.configurableMetadata.numViewsInStakingAuction: "Epoch Views must be greater than staking views"
+                newEpochViews > FlowEpoch.configurableMetadata.numViewsInDKGPhase: "Epoch Views must be greater than DKG phase views"
+            }
+
+            FlowEpoch.configurableMetadata.numViewsInEpoch = newEpochViews
+        }
+
+        pub fun updateAuctionViews(_ newAuctionViews: UInt64) {
+            pre {
+                FlowEpoch.currentEpochPhase == EpochPhase.STAKINGAUCTION: "Can only update fields during the staking auction"
+                newAuctionViews < FlowEpoch.configurableMetadata.numViewsInEpoch: "Epoch Views must be greater than staking views"
+                newAuctionViews > FlowEpoch.configurableMetadata.numViewsInDKGPhase: "Staking views must be greater than DKG phase views"
+            }
+
+            FlowEpoch.configurableMetadata.numViewsInStakingAuction = newAuctionViews
+        }
+
+        pub fun updateDKGPhaseViews(_ newPhaseViews: UInt64) {
+            pre {
+                FlowEpoch.currentEpochPhase == EpochPhase.STAKINGAUCTION: "Can only update fields during the staking auction"
+                newPhaseViews < FlowEpoch.configurableMetadata.numViewsInEpoch: "Epoch Views must be greater than DKG phase views"
+                newPhaseViews < FlowEpoch.configurableMetadata.numViewsInStakingAuction: "Staking views must be greater than DKG phase views"
+            }
+
+            FlowEpoch.configurableMetadata.numViewsInDKGPhase = newPhaseViews
+        }
+
+        pub fun updateNumCollectorClusters(_ newNumClusters: UInt16) {
+            pre {
+                FlowEpoch.currentEpochPhase == EpochPhase.STAKINGAUCTION: "Can only update fields during the staking auction"
+            }
+
+            FlowEpoch.configurableMetadata.numCollectorClusters = newNumClusters
+        }
+    }
+
+    pub resource Heartbeat {
+
+        /// Function that is called every block to advance the epoch
+        pub fun advanceBlock() {
+
+            let currentBlock = getCurrentBlock()
```

Comment:
We might want to re-structure this slightly so we can inject the current block in tests, unless the emulator makes it easy enough to do that.